File: services/boardbox_catalog/src/main/java/com/demoy/boardbox_catalog/service/UserService.java
package com.demoy.boardbox_catalog.service;

import com.demoy.boardbox_catalog.dto.UserDto;
import com.demoy.boardbox_catalog.model.User;
import com.demoy.boardbox_catalog.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    public Mono<UserDto> createUser(UserDto userDto) {
        User user = new User();
        user.setUsername(userDto.username());
        user.setEmail(userDto.email());
        user.setFirstName(userDto.firstName());
        user.setLastName(userDto.lastName());

        return Mono.fromCallable(() -> userRepository.save(user))
                .subscribeOn(Schedulers.boundedElastic())
                .map(this::toDto);
    }

    public Mono<UserDto> getUserById(UUID id) {
        return Mono.fromCallable(() -> userRepository.findById(id))
                .subscribeOn(Schedulers.boundedElastic())
                .flatMap(opt -> opt.isPresent() ? Mono.just(toDto(opt.get())) : Mono.empty());
    }

    public Mono<UserDto> getUserByUsername(String username) {
        return Mono.fromCallable(() -> userRepository.findByUsername(username))
                .subscribeOn(Schedulers.boundedElastic())
                .flatMap(opt -> opt.isPresent() ? Mono.just(toDto(opt.get())) : Mono.empty());
    }

    public Mono<UserDto> updateUser(UUID id, UserDto userDto) {
        return Mono.fromCallable(() -> userRepository.findById(id))
                .subscribeOn(Schedulers.boundedElastic())
                .flatMap(opt -> {
                    if (opt.isPresent()) {
                        User user = opt.get();
                        user.setUsername(userDto.username());
                        user.setEmail(userDto.email());
                        user.setFirstName(userDto.firstName());
                        user.setLastName(userDto.lastName());
                        
                        return Mono.fromCallable(() -> userRepository.save(user))
                                .subscribeOn(Schedulers.boundedElastic())
                                .map(this::toDto);
                    } else {
                        return Mono.empty();
                    }
                });
    }

    public Mono<Void> deleteUser(UUID id) {
        return Mono.fromCallable(() -> {
            userRepository.deleteById(id);
            return null;
        }).subscribeOn(Schedulers.boundedElastic());
    }

    private UserDto toDto(User user) {
        return new UserDto(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getFirstName(),
                user.getLastName(),
                user.getCreatedAt()
        );
    }
}

File: services/order-service/src/main/kotlin/com/demoy/orderservice/service/UserService.kt
package com.demoy.orderservice.service

import com.demoy.orderservice.model.AppUser
import com.demoy.orderservice.repository.UserRepository
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono
import java.util.*

@Service
class UserService(
    private val repo: UserRepository,
    private val template: R2dbcEntityTemplate
) {

    fun createUser(name: String, email: String?): Mono<AppUser> {
        val trimmed = name.trim()
        if (trimmed.isBlank()) {
            return Mono.error(IllegalArgumentException("name is required"))
        }

        val normalizedEmail = email?.trim()?.takeIf { it.isNotBlank() }
        return if (normalizedEmail != null) {
            repo.findByEmail(normalizedEmail)
                .switchIfEmpty(
                    template.insert(AppUser(id = UUID.randomUUID(), name = trimmed, email = normalizedEmail))
                )
        } else {
            template.insert(AppUser(id = UUID.randomUUID(), name = trimmed))
        }
    }

    fun getUser(id: UUID): Mono<AppUser> {
        return repo.findById(id)
            .switchIfEmpty(Mono.error(NoSuchElementException("user not found")))
    }

    fun getUserByEmail(email: String): Mono<AppUser> {
        val normalized = email.trim()
        if (normalized.isBlank()) {
            return Mono.error(IllegalArgumentException("email is required"))
        }
        return repo.findByEmail(normalized)
            .switchIfEmpty(Mono.error(NoSuchElementException("user not found")))
    }

    fun requireUser(id: UUID): Mono<AppUser> = getUser(id)
}
